src/mono/msbuild/WasmAppBuilder/WasmAppBuilder.cs

```diff
@@ -0,0 +1,115 @@
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class WasmAppBuilder : Task
+{
+    // FIXME: Document
+
+    [Required]
+    public string? AppDir { get; set; }
+    [Required]
+    public string? RuntimePackDir { get; set; }
+    [Required]
+    public string? MainAssembly { get; set; }
+    [Required]
+    public string? MainJS { get; set; }
+    [Required]
+    public ITaskItem[]? AssemblySearchPaths { get; set; }
+
+    Dictionary<string, Assembly>? Assemblies;
+    Resolver? Resolver;
+
+    public override bool Execute () {
+        if (!File.Exists (MainAssembly))
+            throw new ArgumentException ($"File MainAssembly='{MainAssembly}' doesn't exist.");
+        if (!File.Exists (MainJS))
+            throw new ArgumentException ($"File MainJS='{MainJS}' doesn't exist.");
+
+        var paths = new List<string> ();
+        Assemblies = new Dictionary<string, Assembly> ();
+
+        // Collect and load assemblies used by the app
+        foreach (var v in AssemblySearchPaths!) {
+            var dir = v.ItemSpec;
+            if (!Directory.Exists (dir))
+                throw new ArgumentException ($"Directory '{dir}' doesn't exist or not a directory.");
+            paths.Add (dir);
+        }
+        Resolver = new Resolver (paths);
+        var mlc = new MetadataLoadContext (Resolver, "System.Private.CoreLib");
+
+        var mainAssembly = mlc.LoadFromAssemblyPath (MainAssembly);
+        Add (mlc, mainAssembly);
+
+        // Create app
+        Directory.CreateDirectory (AppDir!);
+        Directory.CreateDirectory (Path.Join (AppDir, "managed"));
+        foreach (var assembly in Assemblies!.Values)
+            File.Copy (assembly.Location, Path.Join (AppDir, "managed", Path.GetFileName (assembly.Location)), true);
+        foreach (var f in new string [] { "dotnet.wasm", "dotnet.js" })
+            File.Copy (Path.Join (RuntimePackDir, "native", "wasm", "release", f), Path.Join (AppDir, f), true);
+        File.Copy (MainJS!, Path.Join (AppDir, Path.GetFileName (MainJS!)),  true);
+
+        using (var sw = File.CreateText (Path.Join (AppDir, "mono-config.js"))) {
+            sw.WriteLine ("config = {");
+            sw.WriteLine ("\tvfs_prefix: \"managed\",");
+            sw.WriteLine ("\tdeploy_prefix: \"managed\",");
+            sw.WriteLine ("\tenable_debugging: 0,");
+            sw.WriteLine ("\tfile_list: [");
+            foreach (var assembly in Assemblies.Values) {
+                sw.Write ("\"" + Path.GetFileName (assembly.Location) + "\"");
+                sw.Write (", ");
+            }
+            sw.WriteLine ("],");
+            sw.WriteLine ("}");
+        }
+
+        using (var sw = File.CreateText (Path.Join (AppDir, "run-v8.sh"))) {
+            sw.WriteLine ("v8 --expose_wasm runtime.js -- --run " + Path.GetFileName (MainAssembly));
+        }
+
+        return true;
+    }
+
+    void Add (MetadataLoadContext mlc, Assembly assembly) {
```

Comment:
this uses the MetadataLoadContext mainly to figure out which assembly references exist. we don't really need that going forward as we can assume the dotnet toolchain put all the required assemblies in the output folder.

Reply:
we can clean that up later.